### PR TITLE
log cosh dice loss

### PR DESCRIPTION
**Pull request by yumion**
## 変更点
- LogCoshDiceLossを追加
    - dice_loss, tversky_loss, focal_tversky_lossを合わせて追加

## 問題点
- ### `los_cosh_focal_tversky_loss` を使うとlossとiouが `nan` になる
    - focal_tversky_lossもnanになりました

-  dice_loss, tversky_loss, focal_tversky_lossそれぞれ別のfunctional, Lossで作りましたが、既存のlossにただ演算するだけなので、1つにまとめた方がいいですか？

```python
class LogCoshLoss(Loss):
    def __init__(self, base_loss, **kwargs):
        super().__init__(name=f'log_cosh_{base_loss}')
        self.loss = getattr(F, base_loss)
        self.kwargs = kwargs

    def __call__(self, gt, pr):
        x = self.loss(gt, pr, **self.kwargs)
        return tf.math.log((tf.exp(x) + tf.exp(-x)) / 2.0)
```

のように

## reference
https://arxiv.org/abs/2006.14822

### CODE REVIEW

**Comment by aiorhiroki** [comment]
そうしたいけど、
getattrで、呼び出す時に現状はClass名[e.g. LogCoshLoss]でとっていて、
log_cosh_{base_loss}で呼び出せないけど、これで呼び出す方法あるのかな？

**Comment by yumion** [comment]
```yaml
    loss:
      functions:
        LogCoshLoss:
            base_loss: dice_loss
            beta: 1.0
            class_weights: 
                0: 0.0
                1: 1.0
                2: 2.0
                3: 3.0
```
こういう風にyaml書けばいけますけど、出力だとただ `loss` と出て区別つかないのでやめた方がいいですか？

**Comment by yumion** [comment]
```yaml
        LogCoshLoss:
            base_loss: dice_loss
            beta: 1.0
            class_weights: 
                0: 0.0
                1: 1.0
                2: 2.0
                3: 3.0
        LogCoshLoss:
            base_loss: focal_tversky_loss
            alpha: 0.3
            beta: 0.7
            gamma: 1.4
            class_weights: 
                0: 0.0
                1: 1.0
                2: 2.0
                3: 3.0
```
としたらlossがnanになりましたので、分けた今のままで行こうと思います

**Comment by yumion** [comment]
```yaml
      functions:
        CategoricalFocalLoss:
            alpha: 0.25
            gamma: 2.0
            class_weights: 
                0: 0.0
                1: 1.0
                2: 2.0
                3: 3.0
        LogCoshLoss:
            base_loss: dice_loss
            beta: 1.0
            class_weights: 
                0: 0.0
                1: 1.0
                2: 2.0
                3: 3.0
        LogCoshLoss:
            base_loss: tversky_loss
            alpha: 0.3
            beta: 0.7
            class_weights: 
                0: 0.0
                1: 1.0
                2: 2.0
                3: 3.0
```

LogCoshLossを2つ書くとどっちか一方しか採用されてないことがわかったので、classは分けたままにします

**Review by aiorhiroki — APPROVED** [review]
LGTM！！
